src/v/cluster/partition_balancer_planner.cc

```diff
@@ -86,6 +86,19 @@ void partition_balancer_planner::calculate_unavailable_nodes(
     }
 }
 
+bool partition_balancer_planner::all_reports_received(
+  const reallocation_request_state& rrs) {
+    for (const auto& node : _partition_allocator.state().allocation_nodes()) {
```

Comment:
`_partition_allocator.state().allocation_nodes()`

this is classic red flag for use-after-free-bug. in this case `state()` returns a reference which is safe, but if it ever changed to return a value then range-based for loop would hold on to a freed reference (if you look at what range-based for loops expand to internally).

you can grab the reference before the loop, or do something like `for (auto& s = _p.state(); const auto& node : s.nodes())`.

maybe it won't ever return a value, but it's easier to guard against it then it is to find the issue later

Reply:
fixed